### PR TITLE
Refactor endpoint param processing 

### PR DESCRIPTION
**Pull request by Vektrat**
This is a suggestion to see if major changes are required before applying this refactor to the rest of the project

with this approach, each endpoint would not need to be individually tested for params, but instead we could test the parsers, and that `parseParams` is called with the list of param names and parser types the endpoint requires

i've thought of adding an `isRequired` flag for queryParams (not so useful), and most importantly body, but i'm unsure how worth it it is... it would allow this logic to be put into the parseParams side, having generic testing decoupled from the endpoints, but i think required params are not that common in our project, not to mention there'd still be extra checks to be made in many cases (such as mutex params, etc)

### CODE REVIEW

File: rest/src/routes/routeUtils.js [hunk 1]
```diff
@@ -36,9 +36,11 @@ const constants = {
 	}
 };
 
+class ArrayParserError extends Error {}
+
 const isObjectId = str => 24 === str.length && convert.isHexString(str);
 
-const namedParserMap = {
+const parsers = {
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply rest/src/routes/routeUtils.js]: yes i think at this point parsers can be moved to own file

File: rest/src/routes/routeUtils.js [hunk 2]
```diff
@@ -83,10 +85,39 @@ const namedParserMap = {
 			return convert.hexToUint8(str);
 
 		throw Error(`invalid length of hash512 '${str.length}'`);
+	},
+	// available only for body params
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply rest/src/routes/routeUtils.js]: i'm not sure what you're referencing? we've configured restify to merge all params?

[the remark this answers is not in the document]
Comment (Vektrat) [reply rest/src/routes/routeUtils.js]: yep, as explained here https://github.com/nemtech/catapult-rest/pull/59#discussion_r304822737

as per:
http://restify.com/docs/plugins-api/#queryparser (`options.mapParams`)

**Comment by Jaguar0625** [comment]
i think it would be helpful if you could write tests as you envision for one of the changed route functions. i don't see anything principally wrong with the changes, but it's also not clear where the savings are coming from?

1. i imagine you'd want to eventually pass in parsers via services?
2. do we have reuse parameter names for different types. i.e. if `publicKey` is always byte32, you could pass in just name and have name => byte32 mapping in parser class vs passing in transformation (publicKey: convert.hexToUint8)

File: rest/src/routes/accountRoutes.js [hunk 3]
```diff
@@ -25,13 +25,17 @@ const routeResultTypes = require('./routeResultTypes');
...
 const { convert } = catapult.utils;
+const { parsers } = routeUtils;
 
 module.exports = {
 	register: (server, db, services) => {
 		const transactionSender = routeUtils.createSender(routeResultTypes.transaction);
 
 		server.get('/account/:accountId', (req, res, next) => {
-			const [type, accountId] = routeUtils.parseArgument(req.params, 'accountId', 'accountId');
+			const [type, accountId] = routeUtils.parseParams(req.params, {
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply rest/src/routes/accountRoutes.js]: i don't think it's a problem unless you want to restrict certain parameters to certain sources. i'm not sure if we have any such scenario in mind?

**Comment by Jaguar0625** [comment]
relatedly, i think this would be best as (at least) a two part change

1. refactored parser object + tests
2. route integration

File: rest/src/routes/routeUtils.js [hunk 4]
```diff
@@ -83,10 +85,39 @@ const namedParserMap = {
 			return convert.hexToUint8(str);
 
 		throw Error(`invalid length of hash512 '${str.length}'`);
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply rest/src/routes/routeUtils.js]: i think they can be propagated out, if you look at errors.toRestError message is extracted

File: rest/src/routes/accountRoutes.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vektrat) [reply rest/src/routes/accountRoutes.js]: nothing specific, i thought it made the API a bit lazy/not so strict, i just think that each kind of param belongs to a source, and it feels cleaner to have them split, also collisions are a possibility if not careful (it's true though that it is unlikely and that we have none atm - but if it's looking fine now, better, less hassle

**Pull request closed without merging**